File: bench/cat.java
// BOT-GENERATED

import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;

final class cat {
  private static void copy(
    FileChannel input, FileChannel output, ByteBuffer buffer
  ) throws IOException {
    while (input.read(buffer) != -1) {
      buffer.flip();
      while (buffer.hasRemaining()) {
        output.write(buffer);
      }
      buffer.clear();
    }
  }

  public static void main(String[] args) throws IOException {
    var stdin = new FileInputStream(FileDescriptor.in);
    var stdout = new FileOutputStream(FileDescriptor.out);
    var output = stdout.getChannel();

    /*
    Why an explicit 64 KiB loop:
    - `transferTo(System.out)` falls back on JDK's generic 16 KiB loop,
      which takes longer than with our 64 KiB buffer.
    - `transferTo(rawStdout)` can `mmap` regular files. `/dev/null` may
      discard those mappings without faulting the input pages, giving
      fake throughput during the benchmark; that path is also slower
      for real pipe and file outputs.
    - A `byte[]` loop is slower here. A direct buffer lets `FileChannel`
      avoid an intermediate buffer around native I/O.
    */
    var buffer = ByteBuffer.allocateDirect(64 * 1024);

    if (args.length == 0) {
      copy(stdin.getChannel(), output, buffer);
    }

    for (String name : args) {
      if (name.equals("-")) {
        copy(stdin.getChannel(), output, buffer);
        continue;
      }

      try (var input = new FileInputStream(name)) {
        copy(input.getChannel(), output, buffer);
      }
    }

    stdout.flush();
  }
}

File: bench/cat.py
# BOT-GENERATED

import sys


CHUNK_SIZE = 64 * 1024


def copy(source, output):
    # Buffered streams make `shutil.copyfileobj` extremely slow on PyPy here.
    # Raw `FileIO` stays fast; this loop also handles partial writes.
    while chunk := source.read(CHUNK_SIZE):
        remaining = memoryview(chunk)
        while remaining:
            remaining = remaining[output.write(remaining):]


def main():
    stdin = sys.stdin.buffer.raw
    stdout = sys.stdout.buffer.raw

    if not sys.argv[1:]:
        copy(stdin, stdout)
    else:
        for name in sys.argv[1:]:
            if name == "-":
                copy(stdin, stdout)
            else:
                with open(name, "rb", buffering=0) as source:
                    copy(source, stdout)


if __name__ == "__main__":
    main()
